contracts/EthereumLightClient.sol

```diff
@@ -87,12 +92,38 @@ contract EthereumLightClient is Ethash, Initializable, PausableUpgradeable {
         _setFirstBlock(storedBlock);
     }
 
+    function _deleteBlock(uint256 blockNo) private {
+        uint256[] storage blockHashes = blocksByHeight[blockNo];
+        for(uint256 j = 0; j < blockHashes.length; j++) {
+            uint256 blockHash = blockHashes[j];
+            delete blocks[blockHash];
+            delete blockExisting[blockHash];
+        }
+        delete blocksByHeight[blockNo];
+        delete blocksByHeightExisting[blockNo];
+        delete finalizedBlocks[blockNo];
+        delete verifiedBlocks[blockNo];
+    }
+ 
+    function _pruneBlocks(uint256 pruneTime) private {
+        uint256 blockCur = oldestBlockStored;
+        for(uint256 i = 0; i < MAX_PRUNE_ONCE; i++) {
+            uint256 blockNo = blockCur++;
+            if(!blocksByHeightExisting[blockNo]) continue;
+            uint256 blockHash = blocksByHeight[blockNo][0];
+            if(blocks[blockHash].time > pruneTime) break;
+            _deleteBlock(blockNo);
+        }
+        if(blockCur > oldestBlockStored) oldestBlockStored = blockCur;
+    }
+
     //uint32 constant loopAccesses = 64;      // Number of accesses in hashimoto loop
     function addBlockHeader(
         bytes memory _rlpHeader,
         bytes32[4][loopAccesses] memory cache,
         bytes32[][loopAccesses] memory proofs
     ) public whenNotPaused returns (bool) {
+        _pruneBlocks(block.timestamp - BLOCK_EXPIRED);
```

Comment:
What if someone add a bunch of malicious headers (e.g. 30 days) following a fork and cause the contract to prune all existing (valid) blocks?

Reply:
`block.timestamp` is the time of the chain. It does not depend on the submitted block.

Reply:
That's right. Sorry